### PR TITLE
[OP] new mode to reshape

### PR DESCRIPTION
**Pull request by antinucleon**
@piiswrong @winstywang

### CODE REVIEW

**Comment by pluskid** [comment]
In terms of interface for reshape, I always found the [lasagne's way](http://lasagne.readthedocs.io/en/latest/modules/layers/shape.html#lasagne.layers.ReshapeLayer) quite expressive. For example, `([1], [0])` means transpose. `(-1,)` means flat (same for us), and `([0], -1)` means flat everything except the mini-batch.

File: src/operator/reshape-inl.h [hunk 1]
```diff
@@ -25,17 +25,84 @@ enum ReshapeOpInputs {kData};
 enum ReshapeOpOutputs {kOut};
 }  // namespace reshape_enum
 
+
+struct ShapeInfo {
```
Comment (piiswrong) [inline src/operator/reshape-inl.h]: do you need to wrap this? Can you just typedef vector<int> ShapeInfo?

File: src/operator/reshape-inl.h [hunk 2]
```diff
@@ -25,17 +25,84 @@ enum ReshapeOpInputs {kData};
...
+  ShapeInfo shape;
   DMLC_DECLARE_PARAMETER(ReshapeParam) {
     DMLC_DECLARE_FIELD(target_shape)
-    .describe("Target new shape. One and only one dim can be 0, "
+    .set_default(TShape())
+    .describe("(Deprecated!) Target new shape. One and only one dim can be 0, "
               "in which case it will be inferred from the rest of dims");
     DMLC_DECLARE_FIELD(keep_highest).set_default(false)
-    .describe("Whether keep the highest dim unchanged."
+    .describe("(Deprecated!) Whether keep the highest dim unchanged."
```
Comment (piiswrong) [inline src/operator/reshape-inl.h]: (Deprecated!) use shape instead.

**Comment by antinucleon** [comment]
@pluskid Yes, however,  it requires a little more work to resolve input string. How do you think about it? @winstywang @pluskid @tqchen

**Comment by pluskid** [comment]
just to mention that. i am very happy with our reshape current api, too.

File: src/operator/reshape-inl.h (hunk 1 above)
Comment (tqchen) [inline src/operator/reshape-inl.h]: needed for operator overloading

**Comment by antinucleon** [comment]
@pluskid Another way @piiswrong proposed is: use negative number for original dim, eg
```
dshape = (64, 3, 7, 7) # start from 1 to n
oshape = reshape(dshape, (-1, 0, -2, 7)) # 0 means infer,  positive means new dim, negative means original dim
# oshape = (64, 7, 3, 7)
```
I think in this way it is same to lasagne's way in functional, but in a new syntax.

**Comment by pluskid** [comment]
@antinucleon Could be an option, but this could be confusing. Especially Python counts from 0, yet `-1` means the 0-th dimension. It adds confusion when other languages counting from 1 is involved. I am starting to wondering in what scenario do we need to refer to the i-th original dimension. If this is rarely used, we might just go with the current implementation?

**Comment by antinucleon** [comment]
ok, let's merge current way.
On Thu, May 12, 2016 at 18:42 Chiyuan Zhang <notifications@github.com>
wrote:

> @antinucleon <https://github.com/antinucleon> Could be an option, but
> this could be confusing. Especially Python counts from 0, yet -1 means
> the 0-th dimension. It adds confusion when other languages counting from 1
> is involved. I am starting to wondering in what scenario do we need to
> refer to the i-th original dimension. If this is rarely used, we might just
> go with the current implementation?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/dmlc/mxnet/pull/2114#issuecomment-218931652>
>
-- 
Sent from mobile phone

**Pull request merged**